Additional Seccomp tests

mikebrow opened the pull request:
Adding support for seccomp testing. 

Will add pod testing after validating these container tests.

@Random-Liu

Random-Liu commented:
@mikebrow According to https://github.com/kubernetes-incubator/cri-tools/pull/135, seccomp is not supported in the current version of Docker we are using.

mikebrow commented:
Wish I'd known someone was drafting seccomp tests :-)

mikebrow commented:
@Random-Liu @feiskyer all good, I don't see any overlap in our test cases. I'll normalize and combine them on Monday. For example, he's using chmod and grep which should work with docker/default and they do not require caps to be set before generating the default seccomp profile (which from my investigation is different with and without certain caps and on different platforms. I'm using hostname which requires caps to make docker/default allow sethostname, so I'm testing with and without caps on to cover both fail and pass cases.  I'm also testing the privileged case to make sure you can't block a command when privileged, and i've got a test for runtime/default as that may also be used instead of docker/default. And since I did the push I was adding a test for "foo" as a profile to make sure there is a test forcing the profile name to be of the format "localhost/foo"

feiskyer commented:
@mikebrow Sorry for didn't let you know this. But cool with more complete test cases.

feiskyer commented:
@mikebrow needs a rebase now

mikebrow commented:
Rebased. 

@feiskyer  I see there was a change to the CRI api documentation

https://github.com/kubernetes/kubernetes/commit/c242432a3b566ac82d3613689d8af92228cb6f58#diff-df996ad3ce05a000c2009d6e8cfba507L486

Talked to @Random-Liu about this. Will just remove the runtime/default tests for now.

mikebrow commented:
@Random-Liu @feiskyer rebased. Tests passing ready for review.

feiskyer commented on pkg/validate/security_context.go:
```diff
@@ -402,6 +424,41 @@ var _ = framework.KubeDescribe("Security Context", func() {
 		})
+	})
+
+	Context("runtime should support seccomp security context", func() {
```
nits: `[security context]` same with other places.

mikebrow replied:
the top level bucket is "[k8s.io] security context" 
This context is for seccomp tests within [k8s.io] security context.

Not sure what the point is of having brackets to both the left and right most extent of the individual tests.

Let me refactor all of the titles making it easier to regexp and only having brackets on the left for ease of reading.

Random-Liu replied:
@feiskyer On second thought, I also think it is fine to remove the tag given how cri validation test is organized today. :)

mikebrow commented:
Refactored the context container descriptions to fix regexp options and make it easier to read. Removed repetitive text from the suite descriptions again to make it easier to read and easier to identify regexp options.

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -225,7 +247,7 @@ var _ = framework.KubeDescribe("Security Context", func() {
 	})
 
-	Context("runtime should support container with security context", func() {
+	Context("bucket", func() {
```
What do you mean by "bucket"?

mikebrow replied:
a bucket of tests...  This is the second "bucket" of bucket-2, would be another possible name.  This context is already within the "Security Context" .. the other buckets at this level are NamespaceOption and SeccompProfilePath.  

This bucket comprises tests for: SupplementalGroups, RunAsUser, RunAsUserName, ReadOnlyRootfs, Privileged, and Capability. I couldn't come up with a better name for these than the "Security Context bucket"

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -402,8 +424,43 @@ var _ = framework.KubeDescribe("Security Context", func() {
 		})
+	})
+
+	Context("SeccompProfilePath", func() {
```
Probably just `Seccomp`?

mikebrow replied:
I like using the litteral CRI spec name here as it doesn't test all of Seccomp here. It is literally testing the CRI spec defined options for the SeccompProfilePath:		
```&runtimeapi.LinuxPodSandboxConfig{
			SecurityContext: &runtimeapi.LinuxSandboxSecurityContext{
				SeccompProfilePath: string,
			},
		},```

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -37,6 +38,27 @@ import (
 	nginxContainerImage string = "nginx"
+	localhost           string = "localhost/"
+	// profile which denies sethostname syscall
+	seccompBlockHostNameProfile = `{
```
Why not moving those 2 consts into the Seccomp context. No one else is using it.

mikebrow replied:
sure I'll move..

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -402,8 +424,43 @@ var _ = framework.KubeDescribe("Security Context", func() {
+
+		BeforeEach(func() {
+			profileDir, err = createSeccompProfileDir()
+			if err != nil {
```
Don't think we could just log and return, it will still run test then.
You should do assertion here `Expect(err).NotTo(HaveOccurred())`.

Random-Liu replied:
@mikebrow This is the issue we need to fix.

mikebrow replied:
agreed will update ..

feiskyer commented:
@mikebrow thanks 👍

feiskyer commented:
/lgtm

Random-Liu commented:
@feiskyer I'm still reviewing. There are issues. :p

feiskyer commented:
hmm, didn't notice issues. we could open another fix pr for the issues.

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -450,8 +502,123 @@ var _ = framework.KubeDescribe("Security Context", func() {
+			checkSetHostname(rc, containerID, true)
+		})
+
+		It("runtime should block setting hostname with no seccomp profile specified", func() {
```
no seccomp profile means "unconfined", why set hostname will be blocked? By whom?
Default capabilities? If so, I feel like this is more a capability test instead of seccomp test.

Random-Liu replied:
WE can add a test to create a default container, and check whether seccomp is enabled. That is useful.

mikebrow replied:
This is a duplicate test, I'll delete it.

mikebrow commented:
There was an offline discussion about adding another context to separate docker/default tests. I'll push a pr for that in the morn...

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -450,8 +502,123 @@ var _ = framework.KubeDescribe("Security Context", func() {
+			checkSetHostname(rc, containerID, false)
+		})
+
+		It("runtime should support setting hostname with docker/default seccomp profile and sysCaps", func() {
```
Move this to a separate `docker/default` context.

mikebrow replied:
Yes, we both had that idea at the same time :-)

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -450,8 +502,123 @@ var _ = framework.KubeDescribe("Security Context", func() {
+			checkSetHostname(rc, containerID, true)
+		})
+
+		It("runtime should block sethostname with docker/default seccomp profile and no sysCaps", func() {
```
ditto.

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -450,8 +502,123 @@ var _ = framework.KubeDescribe("Security Context", func() {
+			checkSetHostname(rc, containerID, false)
+		})
+
+		It("runtime should block sethostname with nil seccomp profile and no sysCaps", func() {
```
Same. This is a default capability test, not seccomp test.

mikebrow replied:
see above response to your comments.. maybe we should do a quick call on this.

Random-Liu commented on pkg/validate/security_context.go:
```diff
@@ -450,8 +502,123 @@ var _ = framework.KubeDescribe("Security Context", func() {
+			checkSetHostname(rc, containerID, false)
+		})
+
+		It("runtime should block sethostname with unconfined seccomp profile and no sysCaps", func() {
```
Same. This is a default capability test, not seccomp test.

mikebrow replied:
it's both a default caps test and a seccomp test. See unconfined and default rules below:
```
	// Seccomp profile for the sandbox, candidate values are:
	// * docker/default: the default profile for the docker container runtime
	// * unconfined: unconfined profile, ie, no seccomp sandboxing
	// * localhost/<full-path-to-profile>: the profile installed on the node.
	//   <full-path-to-profile> is the full path of the profile.
	// Default: "", which is identical with unconfined.
	SeccompProfilePath string `protobuf:"bytes,7,opt,name=seccomp_profile_path,json=seccompProfilePath,proto3" json:"seccomp_profile_path,omitempty"`
```
If you have an idea for a better way to test the unconfined and nil SeccompProfilePath options.. I'm all ears :)

Random-Liu commented:
@mikebrow Please address the comments.
